src/explore-education-statistics-common/src/modules/table-tool/utils/applyTableHeadersOrder.ts

```diff
@@ -0,0 +1,342 @@
+    return defaultTableHeaders;
+  }
+
+  // Make sure columns is populated by taking the first of the rows groups
+  if (!allOrderedColumnGroupsLength) {
+    return {
+      columnGroups: [],
+      columns: allOrderedRowGroups[0],
+      rowGroups: allOrderedRowGroups.slice(1, -1),
+      rows: allOrderedRowGroups[allOrderedRowGroupsLength - 1],
+    };
+  }
+
+  // Make sure rows is populated by taking the first of the columns groups
+  if (!allOrderedRowGroupsLength) {
+    return {
+      columnGroups: allOrderedColumnGroups.slice(1, -1),
+      columns: allOrderedColumnGroups[allOrderedColumnGroupsLength - 1],
+      rowGroups: [],
+      rows: allOrderedColumnGroups[0],
+    };
+  }
+
+  return {
+    columnGroups: allOrderedColumnGroups.slice(0, -1),
+    columns: allOrderedColumnGroups[allOrderedColumnGroupsLength - 1],
+    rowGroups: allOrderedRowGroups.slice(0, -1),
+    rows: allOrderedRowGroups[allOrderedRowGroupsLength - 1],
+  };
+};
+
+/**
+ * Order the groups for an axis:
+ * - filters the reordered groups to remove those no longer in the headers,
+ * checks both axes
+ * - adds groups that have been added to the headers after the reordered ones
+ * or in default order if groups have not been reordered
+ * - if returnNewGroups is true then the new groups are returned from here
+ * instead of being added to the reordered groups
+ */
+function orderGroups({
+  defaultGroups,
+  reorderedGroups,
+  otherAxisDefaultGroups,
+  otherAxisReorderedGroups,
+  returnNewGroups = false,
+}: {
+  defaultGroups: Filter[][];
+  reorderedGroups: Filter[][];
+  otherAxisDefaultGroups: Filter[][];
+  otherAxisReorderedGroups: Filter[][];
+  returnNewGroups?: boolean;
+}): { orderedGroups: Filter[][]; newGroups: Filter[][] } {
+  const orderedGroups = filterReorderedGroups({
+    reorderedGroups,
+    defaultGroups,
+    otherAxisDefaultGroups,
+  });
+
+  const reorderedGroupsOrder = reorderedGroups.map(group =>
+    getFilterType(group[0]),
+  );
+
+  const defaultGroupsOrder = defaultGroups.map(group =>
+    getFilterType(group[0]),
+  );
+
+  const otherAxisReorderedGroupsOrder = otherAxisReorderedGroups.map(group =>
+    getFilterType(group[0]),
+  );
+
+  const otherAxisDefaultGroupsOrder = otherAxisDefaultGroups.map(group =>
+    getFilterType(group[0]),
+  );
+
+  const groupsAreReordered = areGroupsReordered({
+    defaultOrder: defaultGroupsOrder,
+    otherAxisDefaultOrder: otherAxisDefaultGroupsOrder,
+    previousOrder: reorderedGroupsOrder,
+    previousOtherAxisOrder: otherAxisReorderedGroupsOrder,
+  });
+
+  const newGroups = getNewGroups({
+    otherAxisReorderedGroups,
+    reorderedGroups,
+    defaultGroups,
+  });
+
+  const allGroups = groupsAreReordered
+    ? [...orderedGroups, ...newGroups]
+    : [...orderedGroups, ...newGroups].sort((a, b) => {
+        return (
+          defaultGroupsOrder.indexOf(getFilterType(a[0])) -
+          defaultGroupsOrder.indexOf(getFilterType(b[0]))
+        );
+      });
```

Comment:
This is a little inefficient as we end up creating the `allGroups` regardless of if we actually return them or not. However, we might not need this at all with my new suggestion about returning new groups so that we can fill up the row/col groups more optimally.